### ISSUE TITLE
PPO batch training encountered issues

### ISSUE THREAD

**Issue by Tuziking** [huggingface/trl#2654]
I am using the trl library (v0.11.4) to implement PPO training on llama2. When I set batch_size==1 in PPOConfig, the model output is normal, but when I set batch_size > 1, the model's response appears garbled. Or does the latest version of the TRL library provide a way for me to update to the latest version by customizing simple reward functions instead of reward models? My code is as follow:

import sys
import os
print(sys.path)

import torch
from peft import LoraConfig, TaskType
from transformers import AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig
from trl import AutoModelForCausalLMWithValueHead, PPOConfig, PPOTrainer
from datasets import Dataset
import json
from scripts.utils.calculate_sim import SimilarityCalculator
from scripts.utils.format_catcher import common_catcher
from scripts.utils.format_checker import grammar_checker
from transformers import DataCollatorWithPadding
from torch.cuda.amp import autocast
from tqdm import tqdm
from scripts.utils.my_logger_config import logger
from functools import partial

def make_chat_prompt(user_message):
    return f"""<s>[INST] <<SYS>>
You are a professional json,XML,yaml generation assistant. Please use markdown code blocks to mark the corresponding json, xml, yaml, csv parts.
<</SYS>>

{user_message} [/INST]"""

def load_data(tokenizer):
    file_paths = [
        "samples/xml_true_annotations.json",
        "samples/json_true_annotations.json",
        "samples/yaml_true_annotations.json"
    ]
    all_data = []
    for file_path in file_paths:
        with open(file_path, "r", encoding="utf8") as f:
            data = json.load(f)
            all_data.extend(data)

    dataset = Dataset.from_list(all_data)

    def tokenize_fn(examples):
        prompt = make_chat_prompt(examples["instruction"])
        tokenized = tokenizer(prompt, return_tensors="pt")
        examples["input_ids"] = tokenized["input_ids"][0]
        examples["attention_mask"] = tokenized["attention_mask"][0]
        return examples

    dataset = dataset.map(tokenize_fn, batched=False)
    # dataset.set_format(type='torch', columns=['input_ids', 'attention_mask', 'type', 'instruction', 'answer'])
    return dataset

def my_collate_fn(features, tokenizer):
    if isinstance(features, dict):
        list_of_dicts = [{key: value[i] for key, value in features.items()} for i in range(len(next(iter(features.values()))))]
    elif isinstance(features, list):
        list_of_dicts = features
    else:
        raise TypeError(f"Unsupported type for features: {type(features)}")
    data_collator = DataCollatorWithPadding(tokenizer=tokenizer, return_tensors="pt")
    
    new_features = []
    meta = {
        "type": [],
        "instruction": [],
        "answer": []
    }
    
    for f in list_of_dicts:
        if not isinstance(f, dict):
            raise TypeError(f"Each feature should be a dict, but got {type(f)}")
        if "input_ids" not in f or "attention_mask" not in f:
            raise KeyError("Missing 'input_ids' or 'attention_mask' in feature")
        
        new_features.append({
            "input_ids": f["input_ids"],
            "attention_mask": f["attention_mask"],
        })
        meta["type"].append(f["type"])
        meta["instruction"].append(f["instruction"])
        meta["answer"].append(f["answer"])
    
    batch = data_collator(new_features)
    batch["type"] = meta["type"]
    batch["instruction"] = meta["instruction"]
    batch["answer"] = meta["answer"]
    return batch

if __name__ == "__main__":
    model_path = 'models/Llama-2-7b-chat-hf'
    tokenizer = AutoTokenizer.from_pretrained(model_path, use_fast=False)
    tokenizer.padding_side = "left"
    tokenizer.pad_token = tokenizer.eos_token

    dataset = load_data(tokenizer)
    print('dataset: {}'.format(dataset))
    peft_config = LoraConfig(
        r=8,
        target_modules=["q_proj",
                        "v_proj",
                        "k_proj",
                        "o_proj"
                        ],
        task_type=TaskType.CAUSAL_LM,
        lora_alpha=16,
        lora_dropout=0.1,
        bias="none"
    )

    bnb_config = BitsAndBytesConfig(
        load_in_4bit=True,
        bnb_4bit_use_double_quant=True,
        bnb_4bit_quant_type="nf4",
        bnb_4bit_compute_dtype=torch.bfloat16
    )
    model = AutoModelForCausalLMWithValueHead.from_pretrained(
        model_path,
        quantization_config=bnb_config, 
        device_map="auto",               
        torch_dtype=torch.float16,       
        peft_config=peft_config          
    )

    ppo_config = PPOConfig(kl_penalty="full", batch_size=1, mini_batch_size=1, remove_unused_columns=False)
    ppo_trainer = PPOTrainer(
        config=ppo_config,
        model=model,
        tokenizer=tokenizer,
        dataset=dataset,
        data_collator=partial(my_collate_fn, tokenizer=tokenizer)
    )

    generation_kwargs = {
        "min_length": -1,
        "top_k": 0.0,
        "top_p": 1.0,
        "do_sample": True,
        "pad_token_id": tokenizer.pad_token_id
    }

    sim_calculator = SimilarityCalculator()
    for batch in tqdm(ppo_trainer.dataloader, desc="training progress"):

        query_list = [q for q in batch["input_ids"]]
        answer_tensors = ppo_trainer.generate(
            query_list,  
            return_prompt=False,
            **generation_kwargs
        )

        responses = tokenizer.batch_decode(answer_tensors, skip_special_tokens=True)
        types = batch["type"]
        instructions = batch["instruction"]
        gold_answers = batch["answer"]
        scores = []
        for instruction, response, type_, gold_answer in zip(instructions, responses, types, gold_answers):
            response_catched = common_catcher(response, type_)
            if response_catched is None:
                scores.append(-1)
                continue
            score = sim_calculator.calc_score(target_str = gold_answer, generated_str = response_catched,format_type = type_)
            scores.append(score)
        print('scores: {}'.format(scores))
        rewards = [torch.tensor(x, dtype=torch.float) for x in scores]
        stats = ppo_trainer.step(query_list, answer_tensors, rewards)
    #     ppo_trainer.log_stats(stats, batch, rewards)
    save_path = "models/ppo_trained_model" 
    ppo_trainer.save_model(save_path)
